R/dataframe__frame.R

```diff
@@ -1680,3 +1680,83 @@ DataFrame_sample = function(
   ) |>
     unwrap("in $sample():")
 }
+
+
+
+#' Write to comma-separated values (CSV) file
+#'
+#' @param path File path to which the result should be written.
+#' @param has_header Whether to include header in the CSV output.
+#' @param separator Separate CSV fields with this symbol.
+#' @param line_terminator String used to end each row.
+#' @param quote Byte to use as quoting character.
+#' @param batch_size Number of rows that will be processed per thread.
+#' @param datetime_format A format string, with the specifiers defined by the
+#' chrono Rust crate. If no format specified, the default fractional-second
+#' precision is inferred from the maximum timeunit found in the frame’s Datetime
+#'  cols (if any).
+#' @param date_format A format string, with the specifiers defined by the chrono
+#' Rust crate.
+#' @param time_format A format string, with the specifiers defined by the chrono
+#' Rust crate.
+#' @param float_precision Number of decimal places to write, applied to both
+#' Float32 and Float64 datatypes.
+#' @param null_values A string representing null values (defaulting to the empty
+#' string).
+#' @param quote_style Determines the quoting strategy used.
+#' * `"necessary"` (default): This puts quotes around fields only when necessary.
+#'   They are necessary when fields contain a quote, delimiter or record
+#'   terminator. Quotes are also necessary when writing an empty record (which
+#'   is indistinguishable from a record with one empty field). This is the
+#'   default.
+#' * `"always"`: This puts quotes around every field.
+#' * `"non_numeric"`: This puts quotes around all fields that are non-numeric.
+#'   Namely, when writing a field that does not parse as a valid float or integer,
+#'   then quotes will be used even if they aren`t strictly necessary.
+
+# TODO: include "never" when bumping rust-polars to 0.34
+# * `"never"`: This never puts quotes around fields, even if that results in
+#   invalid CSV data (e.g.: by not quoting strings containing the separator).
+
+#' @return
+#' This doesn't return anything but creates a CSV file.
+#' @export
+#' @rdname IO_write_csv
+#'
+#' @examples
+#' dat = pl$DataFrame(mtcars)
+#'
+
+DataFrame_write_csv = function(
+    path,
+    has_header = TRUE,
+    separator = ",",
+    line_terminator = "\n",
+    quote = '"',
+    batch_size = 1024,
+    datetime_format = NULL,
+    date_format = NULL,
+    time_format = NULL,
+    float_precision = NULL,
+    null_values = "",
+    quote_style = "necessary"
+) {
+
+  if (file_ext(path) != "csv") {
+    stop("Argument `path` must the path to a CSV file.")
+  }
+
+  if (length(quote_style) == 0 ||
+      !quote_style %in% c("always", "necessary", "non_numeric")) {
+    stop("Argument `quote_style` must be one of 'always', 'necessary', or 'non_numeric'.")
+  }
```

Comment:
See https://github.com/pola-rs/r-polars/pull/424#discussion_r1359319634

I think `match.arg` can generally be used for this kind of processing.
However, with the R polars package, the processing is done on the Rust side, so checking on the R side seems not to be necessary.

@sorhawell Thoughts?

Reply:
If I remove this and use `unreachable!` on the rust side, then the error is "Error in .pr$DataFrame$write_csv(self, path, has_header, utf8ToInt(separator), : user function panicked: write_csv"

which I don't find very helpful

Reply:
we could use match.arg but then it should be match.arg(. . . ) |> result() |> uw() to not let a "match.arg" error fly off without context, and that would be all fine. We need to also match.arg on rust side to convert to the underlying enum, and this matching needs error handling too, so we might as well do it there finally. But we can get back to that.

Reply:
unreachable is mostly used to assure the compiler some match arm could not possibly be reached, and rarely tell a developer there is some flawed internal logic here.

```rust
pub fn robj_to_utf8_byte(robj: extendr_api::Robj) -> RResult<u8> {
    let mut utf8_byte_iter = robj_to_str(robj)?.as_bytes().iter();
    match (utf8_byte_iter.next(), utf8_byte_iter.next()) {
        (Some(s), None) => Ok(*s),
        (None, None) => rerr().plain("cannot extract single byte from empty string"),
        (Some(_), Some(_)) => rerr().plain("multi byte-string not allowed"),
        (None, Some(_)) => unreachable!("the iter() cannot yield Some after None(depleted)"),
    }
}
```

here I used unreachable because the compiler could not figure some fact about rust-iterators. Sometimes it can figure these things out, but not here. If one is very hardcore one could use unreachable_unchecked() which becomes "undefined behavior" if actually reached. unreachable_unchecked() is a few nanoseconds faster.

Reply:
Cool, thanks!